# Capstone 2: Data Wrangling (Match Winner Predictor)

## Importing Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Importing Data

### Players File

In [2]:
players_df = pd.read_excel(r'C:/Users/justi/SpringboardDS/Capstone II/data/players.xlsx')
players_df.head()

date player_name           team       opponent        country  \
0 2020-02-26      Brehze  Evil Geniuses         Liquid  United States   
1 2020-02-26        CeRq  Evil Geniuses         Liquid       Bulgaria   
2 2020-02-26       EliGE         Liquid  Evil Geniuses  United States   
3 2020-02-26       Ethan  Evil Geniuses         Liquid  United States   
4 2020-02-26         NAF         Liquid  Evil Geniuses         Canada   

   player_id  match_id  event_id         event_name  best_of  ...  \
0       9136   2339385      4901  IEM Katowice 2020        3  ...   
1      11219   2339385      4901  IEM Katowice 2020        3  ...   
2       8738   2339385      4901  IEM Katowice 2020        3  ...   
3      10671   2339385      4901  IEM Katowice 2020        3  ...   
4       8520   2339385      4901  IEM Katowice 2020        3  ...   

  m3_kddiff_ct m3_adr_ct m3_kast_ct  m3_rating_ct  m3_kills_t  m3_deaths_t  \
0         -1.0      72.5       80.0          0.93         7.0          9.0   
1          3.0      79.5       53.3          1.12         4.0          8.0   
2          1.0      81.5       63.6          1.03         9.0          9.0   
3          0.0      67.2       66.7          0.97         1.0          9.0   
4         -1.0      72.9       81.8          0.96         8.0          7.0   

   m3_kddiff_t  m3_adr_t  m3_kast_t  m3_rating_t  
0         -2.0      70.4       63.6         0.89  
1         -4.0      40.7       54.5         0.53  
2          0.0      87.9       73.3         1.05  
3         -8.0      14.8       45.5         0.31  
4          1.0      56.3       80.0         0.99  

[5 rows x 101 columns]

In [3]:
players_df.shape

(383317, 101)

In [4]:
players_df.dtypes

date           datetime64[ns]
player_name            object
team                   object
opponent               object
country                object
                    ...      
m3_deaths_t           float64
m3_kddiff_t           float64
m3_adr_t              float64
m3_kast_t             float64
m3_rating_t           float64
Length: 101, dtype: object

### Results File

In [5]:
results_df = pd.read_excel(r'C:/Users/justi/SpringboardDS/Capstone II/data/results.xlsx')
results_df.head()

date               team_1          team_2     _map  result_1  \
0 2020-03-18              Recon 5         TeamOne    Dust2         0   
1 2020-03-18              Recon 5         TeamOne  Inferno        13   
2 2020-03-18  New England Whalers        Station7  Inferno        12   
3 2020-03-18              Rugratz  Bad News Bears  Inferno         7   
4 2020-03-18              Rugratz  Bad News Bears  Vertigo         8   

   result_2  map_winner  starting_ct  ct_1  t_2  t_1  ct_2  event_id  \
0        16           2            2     0    1    0    15      5151   
1        16           2            2     8    6    5    10      5151   
2        16           2            1     9    6    3    10      5243   
3        16           2            2     0    8    7     8      5151   
4        16           2            2     4    5    4    11      5151   

   match_id  rank_1  rank_2  map_wins_1  map_wins_2  match_winner  
0   2340454      62      63           0           2             2  
1   2340454      62      63           0           2             2  
2   2340461     140     118          12          16             2  
3   2340453      61      38           0           2             2  
4   2340453      61      38           0           2             2

In [6]:
results_df.shape

(45773, 19)

In [7]:
results_df.dtypes

date            datetime64[ns]
team_1                  object
team_2                  object
_map                    object
result_1                 int64
result_2                 int64
map_winner               int64
starting_ct              int64
ct_1                     int64
t_2                      int64
t_1                      int64
ct_2                     int64
event_id                 int64
match_id                 int64
rank_1                   int64
rank_2                   int64
map_wins_1               int64
map_wins_2               int64
match_winner             int64
dtype: object

## Merging Dataframes and Creating New Columns

In [8]:
# Create a new column 'winning_team' that has the team name that won the match rather than 1 or 2.
mw_list = []
for index, row in results_df.iterrows():
    value = row['match_winner']
    if value == 1:
        mw_list.append(row['team_1'])
    else:
        mw_list.append(row['team_2'])
results_df['winning_team'] = mw_list

In [9]:
# Assure that the new 'winning_team' column has the right values.
results_df.tail()

date team_1      team_2         _map  result_1  result_2  \
45768 2015-11-05     G2  E-frag.net      Inferno        13        16   
45769 2015-11-05     G2  E-frag.net        Dust2        16        13   
45770 2015-11-04    CLG      Liquid      Inferno        16        12   
45771 2015-11-03    NiP    Dignitas        Train        16         4   
45772 2015-11-03    NiP        Envy  Cobblestone        16         9   

       map_winner  starting_ct  ct_1  t_2  t_1  ct_2  event_id  match_id  \
45768           2            1     8    7    5     9      1970   2299059   
45769           1            1    10    5    6     8      1970   2299059   
45770           1            1     7    8    9     4      1934   2299011   
45771           1            2     4    1   12     3      1934   2299001   
45772           1            2     4    6   12     3      1934   2299003   

       rank_1  rank_2  map_wins_1  map_wins_2  match_winner winning_team  
45768       7      16           1           2             2   E-frag.net  
45769       7      16           1           2             2   E-frag.net  
45770      10      14          16          12             1          CLG  
45771       6      12          16           4             1          NiP  
45772       6       1          16           9             1          NiP

In [10]:
# Drop all rows with duplicated match_ids (we only need one map in the match to have the winning_team value b/c it's the same across all maps).
results_df.drop_duplicates(subset = 'match_id', inplace = True, ignore_index = True)

In [11]:
# Assure that the new dataframe has no rows with duplicate match_ids.
results_df['match_id'].is_unique

True

In [12]:
# Drop all columns except match_id and winning_team.
results_df.drop(['date', 'team_1', 'team_2', '_map', 'result_1', 'result_2', 'map_winner', 'starting_ct', 'ct_1', 't_2', 't_1', 'ct_2', 'event_id', 'rank_1', 'rank_2', 'map_wins_1', 'map_wins_2', 'match_winner'], axis=1, inplace=True)

In [13]:
# Assure that the new dataframe has only our two columns of interest in preparation for a merge.
results_df.head()

match_id    winning_team
0   2340454         TeamOne
1   2340461        Station7
2   2340453  Bad News Bears
3   2340456        Endpoint
4   2340455          Espada

In [14]:
# Merge the new results dataframe with the players dataframe on match_id (must do inner join so that we have no null values in the winning_team column).
df = pd.merge(players_df, results_df, how='inner', on='match_id')

In [15]:
# Assure that the new winning_team column has been correctly merged.
df.head()

date player_name           team       opponent        country  \
0 2020-02-26      Brehze  Evil Geniuses         Liquid  United States   
1 2020-02-26        CeRq  Evil Geniuses         Liquid       Bulgaria   
2 2020-02-26       EliGE         Liquid  Evil Geniuses  United States   
3 2020-02-26       Ethan  Evil Geniuses         Liquid  United States   
4 2020-02-26         NAF         Liquid  Evil Geniuses         Canada   

   player_id  match_id  event_id         event_name  best_of  ... m3_adr_ct  \
0       9136   2339385      4901  IEM Katowice 2020        3  ...      72.5   
1      11219   2339385      4901  IEM Katowice 2020        3  ...      79.5   
2       8738   2339385      4901  IEM Katowice 2020        3  ...      81.5   
3      10671   2339385      4901  IEM Katowice 2020        3  ...      67.2   
4       8520   2339385      4901  IEM Katowice 2020        3  ...      72.9   

  m3_kast_ct m3_rating_ct  m3_kills_t  m3_deaths_t  m3_kddiff_t  m3_adr_t  \
0       80.0         0.93         7.0          9.0         -2.0      70.4   
1       53.3         1.12         4.0          8.0         -4.0      40.7   
2       63.6         1.03         9.0          9.0          0.0      87.9   
3       66.7         0.97         1.0          9.0         -8.0      14.8   
4       81.8         0.96         8.0          7.0          1.0      56.3   

   m3_kast_t  m3_rating_t  winning_team  
0       63.6         0.89        Liquid  
1       54.5         0.53        Liquid  
2       73.3         1.05        Liquid  
3       45.5         0.31        Liquid  
4       80.0         0.99        Liquid  

[5 rows x 102 columns]

In [16]:
# Create a new column 'match_outcome' that has a value of Win or Loss based on whether the winning_team and the team they play for match.
o_list = []
for index, row in df.iterrows():
    winner = row['winning_team']
    team = row['team']
    if winner == team:
        o_list.append('Win')
    else:
        o_list.append('Loss')
df['match_outcome'] = o_list

In [17]:
# Assure that the new 'match_outcome' column has the right values.
df.head()

date player_name           team       opponent        country  \
0 2020-02-26      Brehze  Evil Geniuses         Liquid  United States   
1 2020-02-26        CeRq  Evil Geniuses         Liquid       Bulgaria   
2 2020-02-26       EliGE         Liquid  Evil Geniuses  United States   
3 2020-02-26       Ethan  Evil Geniuses         Liquid  United States   
4 2020-02-26         NAF         Liquid  Evil Geniuses         Canada   

   player_id  match_id  event_id         event_name  best_of  ... m3_kast_ct  \
0       9136   2339385      4901  IEM Katowice 2020        3  ...       80.0   
1      11219   2339385      4901  IEM Katowice 2020        3  ...       53.3   
2       8738   2339385      4901  IEM Katowice 2020        3  ...       63.6   
3      10671   2339385      4901  IEM Katowice 2020        3  ...       66.7   
4       8520   2339385      4901  IEM Katowice 2020        3  ...       81.8   

  m3_rating_ct m3_kills_t  m3_deaths_t  m3_kddiff_t  m3_adr_t  m3_kast_t  \
0         0.93        7.0          9.0         -2.0      70.4       63.6   
1         1.12        4.0          8.0         -4.0      40.7       54.5   
2         1.03        9.0          9.0          0.0      87.9       73.3   
3         0.97        1.0          9.0         -8.0      14.8       45.5   
4         0.96        8.0          7.0          1.0      56.3       80.0   

   m3_rating_t  winning_team  match_outcome  
0         0.89        Liquid           Loss  
1         0.53        Liquid           Loss  
2         1.05        Liquid            Win  
3         0.31        Liquid           Loss  
4         0.99        Liquid            Win  

[5 rows x 103 columns]

## Subsetting the best_of Column

In [18]:
# Frequency of bo1, bo2, bo3, and bo5 matches in our data.
df['best_of'].value_counts()

1    133364
3    119630
2     10992
5      1399
Name: best_of, dtype: int64

In [19]:
# Drop best of 2 match rows.
df_filtered = df[df['best_of'] != 2]

In [20]:
# Drop best of 3 match rows.
df_filtered = df_filtered[df_filtered['best_of'] != 3]

In [21]:
# Drop best of 5 match rows.
df_bo1 = df_filtered[df_filtered['best_of'] != 5]

In [22]:
# Assure that bo2, bo3, and bo5 matches are gone.
df_bo1['best_of'].value_counts()

1    133364
Name: best_of, dtype: int64

## Renaming and Dropping Columns

In [23]:
# View all columns in the df.
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
df_bo1.head()

date player_name           team       opponent        country  \
60 2020-02-27    Andersin  Thunder Logic       Station7  United States   
61 2020-02-27    FrostayK       Station7  Thunder Logic  United States   
62 2020-02-27   Inseaniac  Thunder Logic       Station7         Canada   
63 2020-02-27      JonahP       Station7  Thunder Logic         Canada   
64 2020-02-27       PureR  Thunder Logic       Station7  United States   

    player_id  match_id  event_id                        event_name  best_of  \
60      14038   2339816      5151  ESEA MDL Season 33 North America        1   
61      12090   2339816      5151  ESEA MDL Season 33 North America        1   
62      18623   2339816      5151  ESEA MDL Season 33 North America        1   
63      11445   2339816      5151  ESEA MDL Season 33 North America        1   
64      10622   2339816      5151  ESEA MDL Season 33 North America        1   

       map_1 map_2 map_3  kills  assists  deaths  hs  flash_assists  kast  \
60  Overpass   NaN   NaN     21        3      10  12            0.0  88.9   
61  Overpass   NaN   NaN      9        0      16   8            0.0  55.6   
62  Overpass   NaN   NaN     20        4       7   6            0.0  88.9   
63  Overpass   NaN   NaN      5        2      18   2            0.0  38.9   
64  Overpass   NaN   NaN     18        3       9   7            0.0  88.9   

    kddiff    adr  fkdiff  rating  m1_kills  m1_assists  m1_deaths  m1_hs  \
60      11  122.7       4    1.95        21           3         10     12   
61      -7   54.2      -1    0.52         9           0         16      8   
62      13   99.4       4    1.91        20           4          7      6   
63     -13   49.0      -4    0.31         5           2         18      2   
64       9   91.1       2    1.62        18           3          9      7   

    m1_flash_assists  m1_kast  m1_kddiff  m1_adr  m1_fkdiff  m1_rating  \
60               0.0     88.9         11   122.7          4       1.95   
61               0.0     55.6         -7    54.2         -1       0.52   
62               0.0     88.9         13    99.4          4       1.91   
63               0.0     38.9        -13    49.0         -4       0.31   
64               0.0     88.9          9    91.1          2       1.62   

    m2_kills  m2_assists  m2_deaths  m2_hs  m2_flash_assists  m2_kast  \
60       NaN         NaN        NaN    NaN               NaN      NaN   
61       NaN         NaN        NaN    NaN               NaN      NaN   
62       NaN         NaN        NaN    NaN               NaN      NaN   
63       NaN         NaN        NaN    NaN               NaN      NaN   
64       NaN         NaN        NaN    NaN               NaN      NaN   

    m2_kddiff  m2_adr  m2_fkdiff  m2_rating  m3_kills  m3_assists  m3_deaths  \
60        NaN     NaN        NaN        NaN       NaN         NaN        NaN   
61        NaN     NaN        NaN        NaN       NaN         NaN        NaN   
62        NaN     NaN        NaN        NaN       NaN         NaN        NaN   
63        NaN     NaN        NaN        NaN       NaN         NaN        NaN   
64        NaN     NaN        NaN        NaN       NaN         NaN        NaN   

    m3_hs  m3_flash_assists  m3_kast  m3_kddiff  m3_adr  m3_fkdiff  m3_rating  \
60    NaN               NaN      NaN        NaN     NaN        NaN        NaN   
61    NaN               NaN      NaN        NaN     NaN        NaN        NaN   
62    NaN               NaN      NaN        NaN     NaN        NaN        NaN   
63    NaN               NaN      NaN        NaN     NaN        NaN        NaN   
64    NaN               NaN      NaN        NaN     NaN        NaN        NaN   

    kills_ct  deaths_ct  kddiff_ct  adr_ct  kast_ct  rating_ct  kills_t  \
60       3.0        2.0        1.0   145.7    100.0       1.77     18.0   
61       8.0       13.0       -5.0    62.5     60.0       0.60      1.0   
62       5.0        0.0        5.0   192.3    100.0       3.29     15.0   
63   

In [24]:
# Drop columns pertaining to map 2 and map 3.
df_bo1 = df_bo1.drop(['map_2', 'map_3', 'm2_kills', 'm2_assists', 'm2_deaths', 'm2_hs', 'm2_flash_assists', 'm2_kast', 'm2_kddiff', 'm2_adr', 'm2_fkdiff', 'm2_rating', 'm3_kills', 'm3_assists', 'm3_deaths', 'm3_hs', 'm3_flash_assists', 'm3_kast', 'm3_kddiff', 'm3_adr', 'm3_fkdiff', 'm3_rating', 'm2_kills_ct', 'm2_deaths_ct', 'm2_kddiff_ct', 'm2_adr_ct', 'm2_kast_ct', 'm2_rating_ct', 'm2_kills_t', 'm2_deaths_t', 'm2_kddiff_t', 'm2_adr_t', 'm2_kast_t', 'm2_rating_t', 'm3_kills_ct', 'm3_deaths_ct', 'm3_kddiff_ct', 'm3_adr_ct', 'm3_kast_ct', 'm3_rating_ct', 'm3_kills_t', 'm3_deaths_t', 'm3_kddiff_t', 'm3_adr_t', 'm3_kast_t', 'm3_rating_t'], axis=1)

In [25]:
# Assure that the drop was done correctly.
df_bo1.head()

date player_name           team       opponent        country  \
60 2020-02-27    Andersin  Thunder Logic       Station7  United States   
61 2020-02-27    FrostayK       Station7  Thunder Logic  United States   
62 2020-02-27   Inseaniac  Thunder Logic       Station7         Canada   
63 2020-02-27      JonahP       Station7  Thunder Logic         Canada   
64 2020-02-27       PureR  Thunder Logic       Station7  United States   

    player_id  match_id  event_id                        event_name  best_of  \
60      14038   2339816      5151  ESEA MDL Season 33 North America        1   
61      12090   2339816      5151  ESEA MDL Season 33 North America        1   
62      18623   2339816      5151  ESEA MDL Season 33 North America        1   
63      11445   2339816      5151  ESEA MDL Season 33 North America        1   
64      10622   2339816      5151  ESEA MDL Season 33 North America        1   

       map_1  kills  assists  deaths  hs  flash_assists  kast  kddiff    adr  \
60  Overpass     21        3      10  12            0.0  88.9      11  122.7   
61  Overpass      9        0      16   8            0.0  55.6      -7   54.2   
62  Overpass     20        4       7   6            0.0  88.9      13   99.4   
63  Overpass      5        2      18   2            0.0  38.9     -13   49.0   
64  Overpass     18        3       9   7            0.0  88.9       9   91.1   

    fkdiff  rating  m1_kills  m1_assists  m1_deaths  m1_hs  m1_flash_assists  \
60       4    1.95        21           3         10     12               0.0   
61      -1    0.52         9           0         16      8               0.0   
62       4    1.91        20           4          7      6               0.0   
63      -4    0.31         5           2         18      2               0.0   
64       2    1.62        18           3          9      7               0.0   

    m1_kast  m1_kddiff  m1_adr  m1_fkdiff  m1_rating  kills_ct  deaths_ct  \
60     88.9         11   122.7          4       1.95       3.0        2.0   
61     55.6         -7    54.2         -1       0.52       8.0       13.0   
62     88.9         13    99.4          4       1.91       5.0        0.0   
63     38.9        -13    49.0         -4       0.31       NaN        NaN   
64     88.9          9    91.1          2       1.62       4.0        0.0   

    kddiff_ct  adr_ct  kast_ct  rating_ct  kills_t  deaths_t  kddiff_t  adr_t  \
60        1.0   145.7    100.0       1.77     18.0       8.0      10.0  118.1   
61       -5.0    62.5     60.0       0.60      1.0       3.0      -2.0   12.3   
62        5.0   192.3    100.0       3.29     15.0       7.0       8.0   80.8   
63        NaN     NaN      NaN        NaN      NaN       NaN       NaN    NaN   
64        4.0    63.7    100.0       1.74     14.0       9.0       5.0   96.5   

    kast_t  rating_t  m1_kills_ct  m1_deaths_ct  m1_kddiff_ct  m1_adr_ct  \
60    86.7      1.99          3.0           2.0           1.0      145.7   
61    33.3      0.14          8.0          13.0          -5.0       62.5   
62    86.7      1.63          5.0           0.0           5.0      192.3   
63     NaN       NaN          NaN           NaN           NaN        NaN   
64    86.7      1.59          4.0           0.0           4.0       63.7   

    m1_kast_ct  m1_rating_ct  m1_kills_t  m1_deaths_t  m1_kddiff_t  m1_adr_t  \
60       100.0          1.77        18.0          8.0         10.0     118.1   
61        60.0          0.60         1.0          3.0         -2.0      12.3   
62       100.0          3.29        15.0          7.0          8.0      80.8   
63         NaN           NaN         NaN          NaN          NaN       NaN   
64       100.0          1.74        14.0          9.0          5.0      96.5   

    m1_kast_t  m1_rating_t   winning_team match_outcome  
60       86.7         1.99  Thunder Logic           Win  
61       33.3         0.14  Thunder Logic          Loss  
62       86.7         1.63  Thunder Logic           Win  
6

In [26]:
# Drop other columns which store the same information as another (ex. kills = m1_kills), along with a few others.
df_bo1 = df_bo1.drop(['m1_kills', 'm1_assists', 'm1_deaths', 'm1_hs', 'm1_flash_assists', 'm1_kast', 'm1_kddiff', 'm1_adr', 'm1_fkdiff', 'm1_rating', 'm1_kills_ct', 'm1_deaths_ct', 'm1_kddiff_ct', 'm1_adr_ct', 'm1_kast_ct', 'm1_rating_ct', 'm1_kills_t', 'm1_deaths_t', 'm1_kddiff_t', 'm1_adr_t', 'm1_kast_t', 'm1_rating_t', 'winning_team', 'best_of'], axis = 1)

In [27]:
# Rename columns.
df_bo1.rename(columns = {'map_1':'map'}, inplace = True)

In [28]:
# Assure that all of the changes went through.
df_bo1.head()

date player_name           team       opponent        country  \
60 2020-02-27    Andersin  Thunder Logic       Station7  United States   
61 2020-02-27    FrostayK       Station7  Thunder Logic  United States   
62 2020-02-27   Inseaniac  Thunder Logic       Station7         Canada   
63 2020-02-27      JonahP       Station7  Thunder Logic         Canada   
64 2020-02-27       PureR  Thunder Logic       Station7  United States   

    player_id  match_id  event_id                        event_name       map  \
60      14038   2339816      5151  ESEA MDL Season 33 North America  Overpass   
61      12090   2339816      5151  ESEA MDL Season 33 North America  Overpass   
62      18623   2339816      5151  ESEA MDL Season 33 North America  Overpass   
63      11445   2339816      5151  ESEA MDL Season 33 North America  Overpass   
64      10622   2339816      5151  ESEA MDL Season 33 North America  Overpass   

    kills  assists  deaths  hs  flash_assists  kast  kddiff    adr  fkdiff  \
60     21        3      10  12            0.0  88.9      11  122.7       4   
61      9        0      16   8            0.0  55.6      -7   54.2      -1   
62     20        4       7   6            0.0  88.9      13   99.4       4   
63      5        2      18   2            0.0  38.9     -13   49.0      -4   
64     18        3       9   7            0.0  88.9       9   91.1       2   

    rating  kills_ct  deaths_ct  kddiff_ct  adr_ct  kast_ct  rating_ct  \
60    1.95       3.0        2.0        1.0   145.7    100.0       1.77   
61    0.52       8.0       13.0       -5.0    62.5     60.0       0.60   
62    1.91       5.0        0.0        5.0   192.3    100.0       3.29   
63    0.31       NaN        NaN        NaN     NaN      NaN        NaN   
64    1.62       4.0        0.0        4.0    63.7    100.0       1.74   

    kills_t  deaths_t  kddiff_t  adr_t  kast_t  rating_t match_outcome  
60     18.0       8.0      10.0  118.1    86.7      1.99           Win  
61      1.0       3.0      -2.0   12.3    33.3      0.14          Loss  
62     15.0       7.0       8.0   80.8    86.7      1.63           Win  
63      NaN       NaN       NaN    NaN     NaN       NaN          Loss  
64     14.0       9.0       5.0   96.5    86.7      1.59           Win

## Exploring Column Unique Value Counts

In [29]:
# Count of unique values by column.
df_bo1.nunique()

date              1193
player_name       3685
team              1102
opponent          1102
country             82
player_id         3719
match_id         13337
event_id           889
event_name         889
map                  9
kills               62
assists             21
deaths              57
hs                  36
flash_assists       14
kast               338
kddiff              56
adr               1398
fkdiff              24
rating             271
kills_ct            37
deaths_ct           28
kddiff_ct           44
adr_ct            1946
kast_ct            186
rating_ct          403
kills_t             37
deaths_t            35
kddiff_t            44
adr_t             1944
kast_t             201
rating_t           432
match_outcome        2
dtype: int64

The amount of unique values in each columns are about as expected at first glance.

## Data Types and Range of Values

In [30]:
df_bo1.dtypes

date             datetime64[ns]
player_name              object
team                     object
opponent                 object
country                  object
player_id                 int64
match_id                  int64
event_id                  int64
event_name               object
map                      object
kills                     int64
assists                   int64
deaths                    int64
hs                        int64
flash_assists           float64
kast                    float64
kddiff                    int64
adr                     float64
fkdiff                    int64
rating                  float64
kills_ct                float64
deaths_ct               float64
kddiff_ct               float64
adr_ct                  float64
kast_ct                 float64
rating_ct               float64
kills_t                 float64
deaths_t                float64
kddiff_t                float64
adr_t                   float64
kast_t                  float64
rating_t

In [31]:
# Check to see if the flash_assists column is ever not an integer.
df_bo1['flash_assists'].value_counts()

0.0     45454
1.0     26622
2.0     13821
3.0      6604
4.0      3023
5.0      1317
6.0       547
7.0       246
8.0        95
9.0        41
10.0       21
13.0        2
12.0        2
11.0        1
Name: flash_assists, dtype: int64

In [32]:
# Check to see if the kills_ct column is ever not an integer.
df_bo1['kills_ct'].value_counts()

9.0     11148
8.0     11028
10.0    10718
7.0     10441
11.0     9752
6.0      9568
12.0     8744
5.0      8029
13.0     7319
4.0      6850
14.0     6121
3.0      5379
15.0     4763
2.0      4180
16.0     3649
1.0      2866
17.0     2768
18.0     2034
19.0     1375
0.0      1001
20.0      968
21.0      593
22.0      433
23.0      263
24.0      194
25.0      110
26.0       66
27.0       51
28.0       43
29.0       18
30.0       14
32.0        8
31.0        3
33.0        2
37.0        1
41.0        1
34.0        1
Name: kills_ct, dtype: int64

In [33]:
# Check to see if the kills_ct column is ever not an integer.
df_bo1['deaths_ct'].value_counts()

9.0     14812
8.0     14505
10.0    14209
7.0     13250
11.0    12332
6.0     11140
12.0    10034
5.0      8781
13.0     7051
4.0      6829
3.0      4953
14.0     3949
2.0      3268
1.0      1727
15.0     1649
16.0      628
0.0       623
17.0      310
18.0      184
19.0      124
20.0       53
21.0       32
22.0       25
23.0       12
25.0       12
24.0        6
27.0        3
28.0        1
Name: deaths_ct, dtype: int64

In [34]:
# Check to see if the kills_ct column is ever not an integer.
df_bo1['kddiff_ct'].value_counts()

-1.0     10751
 0.0     10647
 1.0     10176
-2.0     10041
 2.0      9324
-3.0      9020
 3.0      8544
 4.0      7482
-4.0      7406
 5.0      6603
-5.0      5779
 6.0      5443
 7.0      4514
-6.0      4296
 8.0      3384
-7.0      2959
 9.0      2772
 10.0     2119
-8.0      1917
 11.0     1515
-9.0      1152
 12.0     1100
 13.0      728
-10.0      699
 14.0      551
 15.0      376
-11.0      318
 16.0      238
-12.0      167
 17.0      149
 18.0      119
 19.0       59
-13.0       49
 20.0       45
 21.0       27
 22.0       10
-14.0        8
 23.0        4
 24.0        4
-16.0        2
-15.0        2
 26.0        1
-20.0        1
 25.0        1
Name: kddiff_ct, dtype: int64

We'll want to convert these columns for both CT and T to have values that are integers once we deal with the null values later on.

In [35]:
df_bo1.describe()

player_id      match_id       event_id         kills  \
count  133364.000000  1.333640e+05  133364.000000  133364.00000   
mean     9004.208872  2.319425e+06    3398.700489      17.64309   
std      4184.916513  1.048589e+04     789.879418       6.32929   
min         2.000000  2.299001e+06    1615.000000       0.00000   
25%      7429.000000  2.310948e+06    2729.000000      13.00000   
50%      9078.000000  2.319395e+06    3374.000000      17.00000   
75%     11586.000000  2.327738e+06    3965.000000      22.00000   
max     20007.000000  2.339828e+06    5224.000000      75.00000   

             assists         deaths             hs  flash_assists  \
count  133364.000000  133364.000000  133364.000000   97796.000000   
mean        4.109137      17.674132       7.984119       1.013917   
std         2.444149       4.811923       3.879718       1.303215   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000      15.000000       5.000000       0.000000   
50%         4.000000      18.000000       8.000000       1.000000   
75%         6.000000      20.000000      10.000000       2.000000   
max        20.000000      57.000000      39.000000      13.000000   

                kast         kddiff            adr         fkdiff  \
count  131805.000000  133364.000000  131800.000000  133364.000000   
mean       69.209672      -0.031043      74.647236      -0.003419   
std        12.338701       7.200305      18.865304       2.437990   
min         6.200000     -26.000000       0.500000     -11.000000   
25%        61.500000      -5.000000      61.700000      -2.000000   
50%        70.000000       0.000000      73.600000       0.000000   
75%        77.800000       5.000000      86.400000       2.000000   
max       100.000000      31.000000     181.000000      12.000000   

              rating       kills_ct      deaths_ct      kddiff_ct  \
count  133364.000000  130502.000000  130502.000000  130502.000000   
mean        1.058114       9.305313       8.407626       0.897687   
std         0.342583       4.624146       3.343237       5.021115   
min         0.050000       0.000000       0.000000     -20.000000   
25%         0.820000       6.000000       6.000000      -3.000000   
50%         1.040000       9.000000       9.000000       1.000000   
75%         1.270000      12.000000      11.000000       4.000000   
max         3.080000      41.000000      28.000000      26.000000   

              adr_ct        kast_ct      rating_ct        kills_t  \
count  130502.000000  130502.000000  130502.000000  130502.000000   
mean       78.371150      71.642680       1.063097       8.423871   
std        26.403034      16.227255       0.457524       4.306291   
min         0.000000       9.100000       0.010000       0.000000   
25%        60.300000      60.000000       0.740000       5.000000   
50%        76.500000      73.300000       1.020000       8.000000   
75%        94.100000      83.300000       1.330000      11.000000   
max       314.000000     100.000000       7.040000      42.000000   

            deaths_t       kddiff_t          adr_t         kast_t  \
count  130502.000000  130502.000000  130502.000000  130502.000000   
mean        9.282670      -0.858799      71.799582      67.435214   
std         3.616436       4.962818      26.620191      16.876572   
min         0.000000     -19.000000       0.000000       6.700000   
25%         7.000000      -4.000000      53.600000      55.600000   
50%        10.000000      -1.000000      69.500000      66.700000   
75%        12.000000       2.000000      87.300000      80.000000   
max        37.000000      28.000000     433.000000     100.000000   

            rating_t  
count  130502.000000  
mean        1.073180  
std         0.485307  
min         0.000000  
25%         0.730000  
50%         1.030000  
75%         1.350000  
max         8.000000

All column value ranges make logical sense. Death columns can have values over 15 because of overtime matches.

## Checking for Duplicates

In [36]:
# Check for unwanted duplicate rows (match_id and player_name).
df_bo1.duplicated(subset=['match_id', 'player_name']).mean()

0.0

## Exploring and Dealing with Null Values

In [37]:
# Check to see which columns, if any, have null values and how many there are.
df_bo1_missing = pd.concat([df_bo1.isnull().sum(), 100 * df_bo1.isnull().mean()], axis=1)
df_bo1_missing.columns=['count', '%']
df_bo1_missing.sort_values(by= ['count', '%'], ascending = False)

count          %
flash_assists  35568  26.669866
kills_ct        2862   2.146006
deaths_ct       2862   2.146006
kddiff_ct       2862   2.146006
adr_ct          2862   2.146006
kast_ct         2862   2.146006
rating_ct       2862   2.146006
kills_t         2862   2.146006
deaths_t        2862   2.146006
kddiff_t        2862   2.146006
adr_t           2862   2.146006
kast_t          2862   2.146006
rating_t        2862   2.146006
adr             1564   1.172730
kast            1559   1.168981
date               0   0.000000
player_name        0   0.000000
team               0   0.000000
opponent           0   0.000000
country            0   0.000000
player_id          0   0.000000
match_id           0   0.000000
event_id           0   0.000000
event_name         0   0.000000
map                0   0.000000
kills              0   0.000000
assists            0   0.000000
deaths             0   0.000000
hs                 0   0.000000
kddiff             0   0.000000
fkdiff             0   0.000000
rating             0   0.000000
match_outcome      0   0.000000

In [38]:
# Replace NULL values with the median value for the flash_assists columnn (using the mean value would result in floats).
df_bo1['flash_assists'].fillna(1, inplace = True)

In [39]:
# Check to see if there is any pattern in which matches have null values.
#df_bo1.loc[df_bo1['kills_ct'].isnull(), 'match_id']

Since there doesn't seem to be any reason or rhyme as to which matches have null values, we should not simply drop the rows when creating/modifying the grouped by team dataframe. If we drop the 2862 rows, then when we later need to aggregate over the grouped match_ids and team names, not all aggregations will have 5 player rows to begin with. For this reason we should impute the missing values instead of dropping the rows like we will for the individual performance dataframe.

In [40]:
# Create a copy of the dataframe so that we can make the grouped by team dataframe using the same structure.
df_bo1_team = df_bo1.copy()

In [41]:
# Drop the NULL rows in the individual player performance dataframe.
df_bo1.dropna(subset = ['kills_ct'], inplace = True)

In [42]:
# Take a close look at the remaining rows with NULL values.
df_bo1.loc[df_bo1['adr'].isnull()]

date player_name              team      opponent        country  \
480    2020-02-25      JMaldo          Monstars    Big Frames  United States   
45647  2019-05-17    Muenster  Party Astronauts  Spacestation  United States   
171214 2017-10-05   autimatic            Cloud9            SK  United States   
254360 2016-04-24      dayV1D             KKona        Liquid  United States   

        player_id  match_id  event_id                             event_name  \
480          5155   2339549      5151       ESEA MDL Season 33 North America   
45647       16681   2333386      4502       ESEA MDL Season 31 North America   
171214       8735   2314373      2867  ESL Pro League Season 6 North America   
254360      11137   2302286      2124                ESL Pro League Season 3   

                map  kills  assists  deaths  hs  flash_assists  kast  kddiff  \
480          Mirage      0        0       2   0            0.0  95.2      -2   
45647      Overpass      0        0       1   0            1.0  96.0      -1   
171214       Mirage      0        0       1   0            0.0  96.0      -1   
254360  Cobblestone      0        0       2   0            1.0  88.2      -2   

        adr  fkdiff  rating  kills_ct  deaths_ct  kddiff_ct  adr_ct  kast_ct  \
480     NaN       0    1.01       0.0        0.0        0.0     0.0    100.0   
45647   NaN       0    0.98       0.0        1.0       -1.0     0.0     90.0   
171214  NaN       0    1.04       0.0        0.0        0.0     0.0    100.0   
254360  NaN       0    0.84       0.0        2.0       -2.0     0.0     86.7   

        rating_ct  kills_t  deaths_t  kddiff_t  adr_t  kast_t  rating_t  \
480          0.98      0.0       2.0      -2.0    0.0    93.3      1.02   
45647        0.85      0.0       0.0       0.0    0.0   100.0      1.07   
171214       1.08      0.0       1.0      -1.0    0.0    93.3      1.01   
254360       0.81      0.0       0.0       0.0    0.0   100.0      1.07   

       match_outcome  
480             Loss  
45647           Loss  
171214          Loss  
254360          Loss

It looks like in these 4 extremely rare cases, both the adr_ct and adr_t values are 0. This would imply that the total adr in these situations were also 0. Since this is such a small subset of our data and because we don't know if this was just simply input incorrectly (adr should have been set to 0 if it was input correctly) we will just drop these 4 rows.

In [43]:
# Check to see if all NULL values are gone.
df_bo1_missing = pd.concat([df_bo1.isnull().sum(), 100 * df_bo1.isnull().mean()], axis=1)
df_bo1_missing.columns=['count', '%']
df_bo1_missing.sort_values(by= ['count', '%'], ascending = False)

count         %
adr                4  0.003065
date               0  0.000000
player_name        0  0.000000
team               0  0.000000
opponent           0  0.000000
country            0  0.000000
player_id          0  0.000000
match_id           0  0.000000
event_id           0  0.000000
event_name         0  0.000000
map                0  0.000000
kills              0  0.000000
assists            0  0.000000
deaths             0  0.000000
hs                 0  0.000000
flash_assists      0  0.000000
kast               0  0.000000
kddiff             0  0.000000
fkdiff             0  0.000000
rating             0  0.000000
kills_ct           0  0.000000
deaths_ct          0  0.000000
kddiff_ct          0  0.000000
adr_ct             0  0.000000
kast_ct            0  0.000000
rating_ct          0  0.000000
kills_t            0  0.000000
deaths_t           0  0.000000
kddiff_t           0  0.000000
adr_t              0  0.000000
kast_t             0  0.000000
rating_t           0  0.000000
match_outcome      0  0.000000

In [44]:
df_bo1.dropna(inplace = True)

There are now no more NULL values in the individual player performance dataframe.

In [45]:
# Now that all of the NULL values are gone, we can change the data types of columns (as mentioned earlier).
df_bo1['flash_assists'] = df_bo1['flash_assists'].astype('int64')

In [46]:
df_bo1['kills_ct'] = df_bo1['kills_ct'].astype('int64')

In [47]:
df_bo1['deaths_ct'] = df_bo1['deaths_ct'].astype('int64')

In [48]:
df_bo1['kddiff_ct'] = df_bo1['kddiff_ct'].astype('int64')

In [49]:
df_bo1['kills_t'] = df_bo1['kills_t'].astype('int64')

In [50]:
df_bo1['deaths_t'] = df_bo1['deaths_t'].astype('int64')

In [51]:
df_bo1['kddiff_t'] = df_bo1['kddiff_t'].astype('int64')

In [52]:
# All data types are now correct.
df_bo1.dtypes

date             datetime64[ns]
player_name              object
team                     object
opponent                 object
country                  object
player_id                 int64
match_id                  int64
event_id                  int64
event_name               object
map                      object
kills                     int64
assists                   int64
deaths                    int64
hs                        int64
flash_assists             int64
kast                    float64
kddiff                    int64
adr                     float64
fkdiff                    int64
rating                  float64
kills_ct                  int64
deaths_ct                 int64
kddiff_ct                 int64
adr_ct                  float64
kast_ct                 float64
rating_ct               float64
kills_t                   int64
deaths_t                  int64
kddiff_t                  int64
adr_t                   float64
kast_t                  float64
rating_t

In [53]:
# Drop the rating columns as they will not be useful in predicting match outcomes later on (rating is determined after a match).
df_bo1 = df_bo1.drop(['rating', 'rating_ct', 'rating_t'], axis = 1)

In [54]:
# Renaming the final dataframe for individual performance and resetting the index.
df_individual_final = df_bo1.reset_index(drop = True)
df_individual_final.head()

date player_name           team       opponent        country  \
0 2020-02-27    Andersin  Thunder Logic       Station7  United States   
1 2020-02-27    FrostayK       Station7  Thunder Logic  United States   
2 2020-02-27   Inseaniac  Thunder Logic       Station7         Canada   
3 2020-02-27       PureR  Thunder Logic       Station7  United States   
4 2020-02-27     Sharkie  Thunder Logic       Station7  United States   

   player_id  match_id  event_id                        event_name       map  \
0      14038   2339816      5151  ESEA MDL Season 33 North America  Overpass   
1      12090   2339816      5151  ESEA MDL Season 33 North America  Overpass   
2      18623   2339816      5151  ESEA MDL Season 33 North America  Overpass   
3      10622   2339816      5151  ESEA MDL Season 33 North America  Overpass   
4      19476   2339816      5151  ESEA MDL Season 33 North America  Overpass   

   kills  assists  deaths  hs  flash_assists  kast  kddiff    adr  fkdiff  \
0     21        3      10  12              0  88.9      11  122.7       4   
1      9        0      16   8              0  55.6      -7   54.2      -1   
2     20        4       7   6              0  88.9      13   99.4       4   
3     18        3       9   7              0  88.9       9   91.1       2   
4     13        5       7   4              0  88.9       6   74.6       2   

   kills_ct  deaths_ct  kddiff_ct  adr_ct  kast_ct  kills_t  deaths_t  \
0         3          2          1   145.7    100.0       18         8   
1         8         13         -5    62.5     60.0        1         3   
2         5          0          5   192.3    100.0       15         7   
3         4          0          4    63.7    100.0       14         9   
4         3          0          3    58.3    100.0       10         7   

   kddiff_t  adr_t  kast_t match_outcome  
0        10  118.1    86.7           Win  
1        -2   12.3    33.3          Loss  
2         8   80.8    86.7           Win  
3         5   96.5    86.7           Win  
4         3   77.9    86.7           Win

## Creating the Team Performance Dataframe

In [55]:
df_bo1_team.head(10)

date player_name           team       opponent        country  \
60 2020-02-27    Andersin  Thunder Logic       Station7  United States   
61 2020-02-27    FrostayK       Station7  Thunder Logic  United States   
62 2020-02-27   Inseaniac  Thunder Logic       Station7         Canada   
63 2020-02-27      JonahP       Station7  Thunder Logic         Canada   
64 2020-02-27       PureR  Thunder Logic       Station7  United States   
65 2020-02-27     Sharkie  Thunder Logic       Station7  United States   
66 2020-02-27      rabbit  Thunder Logic       Station7  United States   
67 2020-02-27       shonk       Station7  Thunder Logic  United States   
68 2020-02-27    sterling       Station7  Thunder Logic         Canada   
69 2020-02-27      zeptic       Station7  Thunder Logic  United States   

    player_id  match_id  event_id                        event_name       map  \
60      14038   2339816      5151  ESEA MDL Season 33 North America  Overpass   
61      12090   2339816      5151  ESEA MDL Season 33 North America  Overpass   
62      18623   2339816      5151  ESEA MDL Season 33 North America  Overpass   
63      11445   2339816      5151  ESEA MDL Season 33 North America  Overpass   
64      10622   2339816      5151  ESEA MDL Season 33 North America  Overpass   
65      19476   2339816      5151  ESEA MDL Season 33 North America  Overpass   
66      19160   2339816      5151  ESEA MDL Season 33 North America  Overpass   
67      14022   2339816      5151  ESEA MDL Season 33 North America  Overpass   
68      14613   2339816      5151  ESEA MDL Season 33 North America  Overpass   
69      14489   2339816      5151  ESEA MDL Season 33 North America  Overpass   

    kills  assists  deaths  hs  flash_assists   kast  kddiff    adr  fkdiff  \
60     21        3      10  12            0.0   88.9      11  122.7       4   
61      9        0      16   8            0.0   55.6      -7   54.2      -1   
62     20        4       7   6            0.0   88.9      13   99.4       4   
63      5        2      18   2            0.0   38.9     -13   49.0      -4   
64     18        3       9   7            0.0   88.9       9   91.1       2   
65     13        5       7   4            0.0   88.9       6   74.6       2   
66     13        3       6   5            0.0  100.0       7   89.2       2   
67      7        0      18   4            0.0   33.3     -11   47.4      -2   
68      8        4      17   4            1.0   55.6      -9   51.8      -5   
69     10        4      16   6            0.0   55.6      -6   71.6      -2   

    rating  kills_ct  deaths_ct  kddiff_ct  adr_ct  kast_ct  rating_ct  \
60    1.95       3.0        2.0        1.0   145.7    100.0       1.77   
61    0.52       8.0       13.0       -5.0    62.5     60.0       0.60   
62    1.91       5.0        0.0        5.0   192.3    100.0       3.29   
63    0.31       NaN        NaN        NaN     NaN      NaN        NaN   
64    1.62       4.0        0.0        4.0    63.7    100.0       1.74   
65    1.48       3.0        0.0        3.0    58.3    100.0       1.80   
66    1.67       0.0        1.0       -1.0    40.0    100.0       1.00   
67    0.32       6.0       15.0       -9.0    47.9     33.3       0.33   
68    0.44       NaN        NaN        NaN     NaN      NaN        NaN   
69    0.63       9.0       13.0       -4.0    69.6     53.3       0.65   

    kills_t  deaths_t  kddiff_t  adr_t  kast_t  rating_t match_outcome  
60     18.0       8.0      10.0  118.1    86.7      1.99           Win  
61      1.0       3.0      -2.0   12.3    33.3      0.14          Loss  
62     15.0       7.0       8.0   80.8    86.7      1.63           Win  
63      NaN       NaN       NaN    NaN     NaN       NaN          Loss  
64     14.0       9.0       5.0   96.5    86.7      1.59           Win  
65     10.0       7.0       3.0   77.9    86.7      1.41           Win  
66     13.0       5.0       8.0   99.1   100.0      1.80           Win  
67      1.0       3.0      -2.0  

In [56]:
# Check to see which columns, if any, have null values and how many there are.
df_bo1_team_missing = pd.concat([df_bo1_team.isnull().sum(), 100 * df_bo1_team.isnull().mean()], axis=1)
df_bo1_team_missing.columns=['count', '%']
df_bo1_team_missing.sort_values(by= ['count', '%'], ascending = False)

count         %
kills_ct        2862  2.146006
deaths_ct       2862  2.146006
kddiff_ct       2862  2.146006
adr_ct          2862  2.146006
kast_ct         2862  2.146006
rating_ct       2862  2.146006
kills_t         2862  2.146006
deaths_t        2862  2.146006
kddiff_t        2862  2.146006
adr_t           2862  2.146006
kast_t          2862  2.146006
rating_t        2862  2.146006
adr             1564  1.172730
kast            1559  1.168981
date               0  0.000000
player_name        0  0.000000
team               0  0.000000
opponent           0  0.000000
country            0  0.000000
player_id          0  0.000000
match_id           0  0.000000
event_id           0  0.000000
event_name         0  0.000000
map                0  0.000000
kills              0  0.000000
assists            0  0.000000
deaths             0  0.000000
hs                 0  0.000000
flash_assists      0  0.000000
kddiff             0  0.000000
fkdiff             0  0.000000
rating             0  0.000000
match_outcome      0  0.000000

In [57]:
# View the summary statistics for each column.
df_bo1_team.describe()

player_id      match_id       event_id         kills  \
count  133364.000000  1.333640e+05  133364.000000  133364.00000   
mean     9004.208872  2.319425e+06    3398.700489      17.64309   
std      4184.916513  1.048589e+04     789.879418       6.32929   
min         2.000000  2.299001e+06    1615.000000       0.00000   
25%      7429.000000  2.310948e+06    2729.000000      13.00000   
50%      9078.000000  2.319395e+06    3374.000000      17.00000   
75%     11586.000000  2.327738e+06    3965.000000      22.00000   
max     20007.000000  2.339828e+06    5224.000000      75.00000   

             assists         deaths             hs  flash_assists  \
count  133364.000000  133364.000000  133364.000000  133364.000000   
mean        4.109137      17.674132       7.984119       1.010205   
std         2.444149       4.811923       3.879718       1.115998   
min         0.000000       0.000000       0.000000       0.000000   
25%         2.000000      15.000000       5.000000       0.000000   
50%         4.000000      18.000000       8.000000       1.000000   
75%         6.000000      20.000000      10.000000       1.000000   
max        20.000000      57.000000      39.000000      13.000000   

                kast         kddiff            adr         fkdiff  \
count  131805.000000  133364.000000  131800.000000  133364.000000   
mean       69.209672      -0.031043      74.647236      -0.003419   
std        12.338701       7.200305      18.865304       2.437990   
min         6.200000     -26.000000       0.500000     -11.000000   
25%        61.500000      -5.000000      61.700000      -2.000000   
50%        70.000000       0.000000      73.600000       0.000000   
75%        77.800000       5.000000      86.400000       2.000000   
max       100.000000      31.000000     181.000000      12.000000   

              rating       kills_ct      deaths_ct      kddiff_ct  \
count  133364.000000  130502.000000  130502.000000  130502.000000   
mean        1.058114       9.305313       8.407626       0.897687   
std         0.342583       4.624146       3.343237       5.021115   
min         0.050000       0.000000       0.000000     -20.000000   
25%         0.820000       6.000000       6.000000      -3.000000   
50%         1.040000       9.000000       9.000000       1.000000   
75%         1.270000      12.000000      11.000000       4.000000   
max         3.080000      41.000000      28.000000      26.000000   

              adr_ct        kast_ct      rating_ct        kills_t  \
count  130502.000000  130502.000000  130502.000000  130502.000000   
mean       78.371150      71.642680       1.063097       8.423871   
std        26.403034      16.227255       0.457524       4.306291   
min         0.000000       9.100000       0.010000       0.000000   
25%        60.300000      60.000000       0.740000       5.000000   
50%        76.500000      73.300000       1.020000       8.000000   
75%        94.100000      83.300000       1.330000      11.000000   
max       314.000000     100.000000       7.040000      42.000000   

            deaths_t       kddiff_t          adr_t         kast_t  \
count  130502.000000  130502.000000  130502.000000  130502.000000   
mean        9.282670      -0.858799      71.799582      67.435214   
std         3.616436       4.962818      26.620191      16.876572   
min         0.000000     -19.000000       0.000000       6.700000   
25%         7.000000      -4.000000      53.600000      55.600000   
50%        10.000000      -1.000000      69.500000      66.700000   
75%        12.000000       2.000000      87.300000      80.000000   
max        37.000000      28.000000     433.000000     100.000000   

            rating_t  
count  130502.000000  
mean        1.073180  
std         0.485307  
min         0.000000  
25%         0.730000  
50%         1.030000  
75%         1.350000  
max         8.000000

In [58]:
# Drop numerical columns that we do not want to aggregate over (id columns other than match_id).
df_bo1_team.drop(columns = ['player_id', 'event_id'], inplace = True)

In [59]:
# Check out the NULL value rows.
df_bo1_team[df_bo1_team['kills_ct'].isnull()].head()

date   player_name      team       opponent      country  match_id  \
63  2020-02-27        JonahP  Station7  Thunder Logic       Canada   2339816   
68  2020-02-27      sterling  Station7  Thunder Logic       Canada   2339816   
210 2020-02-26          Doom     PC419          ORDER    Australia   2339773   
212 2020-02-26  Fluent Swede     PC419          ORDER    Australia   2339773   
214 2020-02-26        Magnum     PC419          ORDER  New Zealand   2339773   

                           event_name       map  kills  assists  deaths  hs  \
63   ESEA MDL Season 33 North America  Overpass      5        2      18   2   
68   ESEA MDL Season 33 North America  Overpass      8        4      17   4   
210      ESEA MDL Season 33 Australia     Dust2      4        1      17   3   
212      ESEA MDL Season 33 Australia     Dust2      4        0      17   3   
214      ESEA MDL Season 33 Australia     Dust2      4        1      17   2   

     flash_assists  kast  kddiff   adr  fkdiff  rating  kills_ct  deaths_ct  \
63             0.0  38.9     -13  49.0      -4    0.31       NaN        NaN   
68             1.0  55.6      -9  51.8      -5    0.44       NaN        NaN   
210            1.0  26.3     -13  28.7      -6    0.25       NaN        NaN   
212            0.0  21.1     -13  35.4      -1    0.30       NaN        NaN   
214            1.0  31.6     -13  31.4       0    0.26       NaN        NaN   

     kddiff_ct  adr_ct  kast_ct  rating_ct  kills_t  deaths_t  kddiff_t  \
63         NaN     NaN      NaN        NaN      NaN       NaN       NaN   
68         NaN     NaN      NaN        NaN      NaN       NaN       NaN   
210        NaN     NaN      NaN        NaN      NaN       NaN       NaN   
212        NaN     NaN      NaN        NaN      NaN       NaN       NaN   
214        NaN     NaN      NaN        NaN      NaN       NaN       NaN   

     adr_t  kast_t  rating_t match_outcome  
63     NaN     NaN       NaN          Loss  
68     NaN     NaN       NaN          Loss  
210    NaN     NaN       NaN          Loss  
212    NaN     NaN       NaN          Loss  
214    NaN     NaN       NaN          Loss

In [60]:
# Ignore this cell --> Old method of dealing with NULL values. We dropped them instead.

# Filling NULL values based on related total columns that have no NULL values.

# The following lines of code use integer division to split a column into values for the related CT and T columns.
#df_bo1_team['kills_ct'] = df_bo1_team['kills_ct'].fillna(df_bo1_team['kills'] // 2)
#df_bo1_team['kills_t'] = df_bo1_team['kills_t'].fillna(df_bo1_team['kills'] // 2)
#df_bo1_team['deaths_ct'] = df_bo1_team['deaths_ct'].fillna(df_bo1_team['deaths'] // 2)
#df_bo1_team['deaths_t'] = df_bo1_team['deaths_t'].fillna(df_bo1_team['deaths'] // 2)
#df_bo1_team['kddiff_ct'] = df_bo1_team['kddiff_ct'].fillna(df_bo1_team['kddiff'] // 2)
#df_bo1_team['kddiff_t'] = df_bo1_team['kddiff_t'].fillna(df_bo1_team['kddiff'] // 2)

# The following lines of code copy the value of a column to the related CT and T columns.
#df_bo1_team['rating_ct'] = df_bo1_team['rating_ct'].fillna(df_bo1_team['rating'])
#df_bo1_team['rating_t'] = df_bo1_team['rating_t'].fillna(df_bo1_team['rating'])
#df_bo1_team['kast_ct'] = df_bo1_team['kast_ct'].fillna(df_bo1_team['kast'])
#df_bo1_team['kast_t'] = df_bo1_team['kast_t'].fillna(df_bo1_team['kast'])
#df_bo1_team['adr_ct'] = df_bo1_team['adr_ct'].fillna(df_bo1_team['adr'])
#df_bo1_team['adr_t'] = df_bo1_team['adr_t'].fillna(df_bo1_team['adr'])

# The following lines of code set the adr to 0 for the 4 rows found earlier to have an adr_ct and adr_t of 0 (we do this because we do not want to impute using the mean for these rows).
#df_bo1_team.loc[(df_bo1_team['adr_t'] == 0) & (df_bo1_team['adr_ct'] == 0), 'adr'] = 0

# The following lines of code copy the average kast and adr (CT and T too), respectively, to rows where all 3 related variables are NULL (i.e. we don't have the total value to copy over).
#df_bo1_team['kast'] = df_bo1_team['kast'].fillna(df_bo1_team['kast'].mean())
#df_bo1_team['kast_ct'] = df_bo1_team['kast_ct'].fillna(df_bo1_team['kast_ct'].mean())
#df_bo1_team['kast_t'] = df_bo1_team['kast_t'].fillna(df_bo1_team['kast_t'].mean())
#df_bo1_team['adr'] = df_bo1_team['adr'].fillna(df_bo1_team['adr'].mean())
#df_bo1_team['adr_ct'] = df_bo1_team['adr_ct'].fillna(df_bo1_team['adr_ct'].mean())
#df_bo1_team['adr_t'] = df_bo1_team['adr_t'].fillna(df_bo1_team['adr_t'].mean())

We will drop all null values since the rows make up a very small subset of our data (2.15%).

In [61]:
# Drop all null values
df_bo1_team.dropna(inplace = True)

In [62]:
df_bo1_team.shape

(130498, 31)

In [63]:
# Check to see if all NULL values are gone.
df_bo1_team_missing = pd.concat([df_bo1_team.isnull().sum(), 100 * df_bo1_team.isnull().mean()], axis=1)
df_bo1_team_missing.columns=['count', '%']
df_bo1_team_missing.sort_values(by= ['count', '%'], ascending = False)

count    %
date               0  0.0
player_name        0  0.0
team               0  0.0
opponent           0  0.0
country            0  0.0
match_id           0  0.0
event_name         0  0.0
map                0  0.0
kills              0  0.0
assists            0  0.0
deaths             0  0.0
hs                 0  0.0
flash_assists      0  0.0
kast               0  0.0
kddiff             0  0.0
adr                0  0.0
fkdiff             0  0.0
rating             0  0.0
kills_ct           0  0.0
deaths_ct          0  0.0
kddiff_ct          0  0.0
adr_ct             0  0.0
kast_ct            0  0.0
rating_ct          0  0.0
kills_t            0  0.0
deaths_t           0  0.0
kddiff_t           0  0.0
adr_t              0  0.0
kast_t             0  0.0
rating_t           0  0.0
match_outcome      0  0.0

In [64]:
# View the data types before we merge.
df_bo1_team.dtypes

date             datetime64[ns]
player_name              object
team                     object
opponent                 object
country                  object
match_id                  int64
event_name               object
map                      object
kills                     int64
assists                   int64
deaths                    int64
hs                        int64
flash_assists           float64
kast                    float64
kddiff                    int64
adr                     float64
fkdiff                    int64
rating                  float64
kills_ct                float64
deaths_ct               float64
kddiff_ct               float64
adr_ct                  float64
kast_ct                 float64
rating_ct               float64
kills_t                 float64
deaths_t                float64
kddiff_t                float64
adr_t                   float64
kast_t                  float64
rating_t                float64
match_outcome            object
dtype: o

In [65]:
# Perform the groupby now that there are no NULL values and therefore we can aggregate over the numerical columns.
df_team_final = df_bo1_team.groupby(['match_id', 'team']).agg({'date':'max', 'opponent':'max', 'event_name':'max', 'map':'max', 'kills':'mean', 'assists':'mean', 'deaths':'mean', 'hs':'mean', 'flash_assists':'mean', 'kast':'mean', 'kddiff':'mean', 'adr':'mean', 'fkdiff':'mean', 'rating':'mean', 'kills_ct':'mean', 'deaths_ct':'mean', 'kddiff_ct':'mean', 'adr_ct':'mean', 'kast_ct':'mean', 'rating_ct':'mean', 'kills_t':'mean', 'deaths_t':'mean', 'kddiff_t':'mean', 'adr_t':'mean', 'kast_t':'mean', 'rating_t':'mean', 'match_outcome':'max',})

In [66]:
# Drop the rating columns as they will not be useful in predicting match outcomes later on (rating is determined after a match).
df_team_final = df_team_final.drop(['rating', 'rating_ct', 'rating_t'], axis = 1)

In [67]:
# Make a copy of the dataframe with the indexes reset.
df_team_final_ri = df_team_final.reset_index()

In [68]:
# View our final dataframe(s).
df_team_final.tail(6)

date       opponent  \
match_id team                                      
2339816  Station7      2020-02-27  Thunder Logic   
         Thunder Logic 2020-02-27       Station7   
2339817  Divine        2020-02-27        Oceanus   
         Oceanus       2020-02-27         Divine   
2339828  Airborne      2020-02-26         Chiefs   
         Chiefs        2020-02-26       Airborne   

                                              event_name       map      kills  \
match_id team                                                                   
2339816  Station7       ESEA MDL Season 33 North America  Overpass   8.666667   
         Thunder Logic  ESEA MDL Season 33 North America  Overpass  17.000000   
2339817  Divine         ESEA MDL Season 33 North America     Dust2  20.200000   
         Oceanus        ESEA MDL Season 33 North America     Dust2  17.600000   
2339828  Airborne           ESEA MDL Season 33 Australia   Inferno  13.200000   
         Chiefs             ESEA MDL Season 33 Australia   Inferno  18.800000   

                         assists     deaths   hs  flash_assists       kast  \
match_id team                                                                
2339816  Station7       1.333333  16.666667  6.0            0.0  48.166667   
         Thunder Logic  3.600000   7.800000  6.8            0.0  91.120000   
2339817  Divine         7.400000  17.800000  9.6            2.4  76.320000   
         Oceanus        5.000000  20.200000  9.0            1.8  68.920000   
2339828  Airborne       2.000000  18.800000  5.2            0.2  56.520000   
         Chiefs         6.000000  13.200000  7.4            2.0  80.880000   

                        kddiff        adr    fkdiff   kills_ct  deaths_ct  \
match_id team                                                               
2339816  Station7         -8.0  57.733333 -1.666667   7.666667  13.666667   
         Thunder Logic     9.2  95.400000  2.800000   3.000000   0.600000   
2339817  Divine            2.4  79.100000  0.200000  12.800000   8.800000   
         Oceanus          -2.6  72.620000 -0.200000   9.000000   7.400000   
2339828  Airborne         -5.6  64.300000 -3.000000   4.000000   7.000000   
         Chiefs            5.6  86.180000  3.000000  11.800000   9.200000   

                        kddiff_ct  adr_ct     kast_ct  kills_t  deaths_t  \
match_id team                                                              
2339816  Station7            -6.0   60.00   48.866667      1.0       3.0   
         Thunder Logic        2.4  100.00  100.000000     14.0       7.2   
2339817  Divine               4.0   87.44   79.980000      7.4       9.0   
         Oceanus              1.6   78.14   71.640000      8.6      12.8   
2339828  Airborne            -3.0   57.56   50.000000      9.2      11.8   
         Chiefs               2.6   83.98   81.340000      7.0       4.0   

                        kddiff_t      adr_t     kast_t match_outcome  
match_id team                                                         
2339816  Station7           -2.0  46.333333  44.433333          Loss  
         Thunder Logic       6.8  94.480000  89.360000           Win  
2339817  Divine             -1.6  68.740000  71.660000           Win  
         Oceanus            -4.2  68.240000  66.680000          Loss  
2339828  Airborne           -2.6  67.860000  60.000000          Loss  
         Chiefs              3.0  90.300000  80.000000           Win

In [69]:
# View our final dataframe(s).
df_team_final_ri.head()

match_id           team       date       opponent  \
0   2300412  Natus Vincere 2016-01-22             SK   
1   2300412             SK 2016-01-22  Natus Vincere   
2   2300413           FaZe 2016-01-22     Luminosity   
3   2300413     Luminosity 2016-01-22           FaZe   
4   2300414       Astralis 2016-01-22       Dignitas   

                          event_name       map  kills  assists  deaths   hs  \
0  DreamHack ZOWIE Open Leipzig 2016    Mirage   15.0      3.2     9.2  5.8   
1  DreamHack ZOWIE Open Leipzig 2016    Mirage    9.2      2.6    15.0  4.6   
2  DreamHack ZOWIE Open Leipzig 2016   Inferno   19.8      4.0    14.6  9.0   
3  DreamHack ZOWIE Open Leipzig 2016   Inferno   14.6      3.6    20.0  5.8   
4  DreamHack ZOWIE Open Leipzig 2016  Overpass   18.2      2.4    13.0  7.2   

   flash_assists   kast  kddiff    adr  fkdiff  kills_ct  deaths_ct  \
0            1.0  74.74     5.8  83.26     1.0       3.4        2.2   
1            1.0  57.90    -5.8  61.58    -1.0       7.0       11.6   
2            1.0  71.20     5.2  84.74     0.2       7.2        6.6   
3            1.0  60.00    -5.4  71.34    -0.2       8.0       12.6   
4            1.0  73.60     5.2  83.08     1.8       6.4        6.6   

   kddiff_ct  adr_ct  kast_ct  kills_t  deaths_t  kddiff_t  adr_t  kast_t  \
0        1.2   89.26    70.00     11.6       7.0       4.6  81.66   76.00   
1       -4.6   59.90    58.68      2.2       3.4      -1.2  67.88   55.00   
2        0.6   79.40    66.00     12.6       8.0       4.6  88.30   74.66   
3       -4.6   69.12    52.00      6.6       7.4      -0.8  74.70   72.00   
4       -0.2   77.38    66.00     11.8       6.4       5.4  86.86   78.66   

  match_outcome  
0           Win  
1          Loss  
2           Win  
3          Loss  
4           Win

In [70]:
# View the data types post-merge.
df_team_final_ri.dtypes

match_id                  int64
team                     object
date             datetime64[ns]
opponent                 object
event_name               object
map                      object
kills                   float64
assists                 float64
deaths                  float64
hs                      float64
flash_assists           float64
kast                    float64
kddiff                  float64
adr                     float64
fkdiff                  float64
kills_ct                float64
deaths_ct               float64
kddiff_ct               float64
adr_ct                  float64
kast_ct                 float64
kills_t                 float64
deaths_t                float64
kddiff_t                float64
adr_t                   float64
kast_t                  float64
match_outcome            object
dtype: object

## Saving the Data

In [71]:
# Save the individual dataframe.
df_individual_final.to_csv('C:/Users/justi/SpringboardDS/Capstone II/1.Match Winner Predictor/df_individual.csv')

In [72]:
# Save the team dataframe.
df_team_final.to_csv('C:/Users/justi/SpringboardDS/Capstone II/1.Match Winner Predictor/df_team.csv')